packages/server/src/manifest-builder.ts

```diff
@@ -77,59 +76,66 @@ function* processManifest(options: ManifestBuilderOptions): Operation {
   let manifest = yield import(distPath);
 
   manifest = manifest.default || manifest;
-
   manifest.fileName = fileName;
 
-
   let slice = options.atom.slice('manifest');
-  slice.set(manifest as Manifest);
+  
+  slice.update(() => ({ ...manifest }));
 
   return distPath;
 }
 
-function logBuildError(error: BundlerError) {
-  console.error("[manifest builder] build error:", error.message);
-  if (error.frame) {
-    console.error("[manifest builder] build error frame:\n", error.frame);
-  }
-}
-
-function* waitForSuccessfulBuild(bundlerEvents: ChainableSubscription<BundlerMessage, undefined>, delegate: Mailbox): Operation {
-  let message: BundlerMessage = yield bundlerEvents.expect();
-
-  if (message.type === "error") {
-    logBuildError(message.error);
-    delegate.send({ event: "error" });
-    yield waitForSuccessfulBuild(bundlerEvents, delegate);
-  }
-}
-
 export function* createManifestBuilder(options: ManifestBuilderOptions): Operation {
+  let bundlerSlice = options.atom.slice('bundler');
+
+  bundlerSlice.set({ type: 'UNBUNDLED' });
+  
   let bundler: Bundler = yield Bundler.create(
     [{
       entry: options.srcPath,
       globalName: bigtestGlobals.manifestProperty,
       outFile: path.join(options.buildDir, "manifest.js")
-    }]
+    }],
   );
-  let bundlerEvents: ChainableSubscription<BundlerMessage, undefined> = yield subscribe(bundler);
-
-  yield waitForSuccessfulBuild(bundlerEvents, options.delegate);
 
-  let distPath: string = yield processManifest(options);
-
-  console.debug("[manifest builder] manifest ready");
-
-  options.delegate.send({ status: "ready", path: distPath });
-
-  yield bundlerEvents.forEach(function*(message) {
-    if (message.type === 'error') {
-      logBuildError(message.error);
-      options.delegate.send({ event: "error" });
-    } else {
-      let distPath = yield processManifest(options);
-      console.info("[manifest builder] manifest updated");
-      options.delegate.send({ event: "update", path: distPath });
+  yield Subscribable.from(bundler).forEach(function* (message) {
+    switch (message.type) {
+      case 'START':
+        console.debug("[manifest builder] received bundler start");
+        
+        bundlerSlice.update(() => ({ type: 'BUILDING', warnings: [] }));
+        break;
+      case 'UPDATE':
+        console.debug("[manifest builder] received bundle update");
+        
+        let path: string = yield processManifest(options);
+        
+        bundlerSlice.update((previous) => {
+          // as this is a typescript assertion it does more than just check the condition
+          // it will type narrow the discriminated union based on the type discriminator
+          assert(previous.type === 'BUILDING', `invalid transition from ${previous.type} to 'GREEN'`);
```

Comment:
These type-narrowing assertions are really cool! However, they will cause the bundler, and the entire orchestrator to crash, we should indicate clearly that this is a bug in BigTest and should be reported... e.g.

```ts
assert(previous.type === 'BUILDING', `
INTERNAL ASSERTION FAILURE:
While bundling test files, an invalid state transition from ${previous.type} to 'GREEN' was received.
this is almost certainly a bug in BigTest, and we would be very grateful if you would report
it at https://github.com/thefrontside/bigtest/issues/new`
`);

Reply:
@cowboyd we could have a template for this type of thing or do we have one?  These errors are definitely for us and of course, should never happen but .....things do happen

Reply:
@cowboyd I might have missed the point of the assertions, I think having a generic one might actually be defeating the purpose.  I think they should actually be for specific purposes.

Let me try another commit.

Reply:
@cowboyd I love the type narrowing and assertions that actually live in the code.  This is where we get into the type system enforcing correctness and not just for plain old checking you can't pass a string to a function that expects a number

Reply:
@cowboyd we could also use rollup to strip the assertions for the prod build

Reply:
@cowboyd I have added more specialised assertions in `@bigtest/server`.  

If this is ok then I will remove the generic one and add a couple of tests.

Reply:
`asserts current is R`

🤯 🤯 🤯 

@dagda1 Sounds like a plan!